File: frontend/src/services/analyticsService.ts
import { apiClient } from './axiosConfig';

export interface TopResource {
  resourceId: number;
  resourceName: string;
  resourceType: string;
  location: string;
  bookingCount: number;
}

export interface PeakHour {
  hour: number;
  bookingCount: number;
  timeLabel: string;
}

export interface BookingsByType {
  type: string;
  bookingCount: number;
}

const ANALYTICS_API = '/api/admin/analytics';

export const analyticsService = {
  /**
   * Get top resources by booking count
   */
  getTopResources: async (limit: number = 10): Promise<TopResource[]> => {
    try {
      const response = await apiClient.get<TopResource[]>(`${ANALYTICS_API}/top-resources`, {
        params: { limit }
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching top resources:', error);
      throw error;
    }
  },

  /**
   * Get top resources by date range
   */
  getTopResourcesByDateRange: async (
    startDate: string,
    endDate: string,
    limit: number = 10
  ): Promise<TopResource[]> => {
    try {
      const response = await apiClient.get<TopResource[]>(`${ANALYTICS_API}/top-resources/range`, {
        params: { startDate, endDate, limit }
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching top resources by date range:', error);
      throw error;
    }
  },

  /**
   * Get peak booking hours
   */
  getPeakBookingHours: async (): Promise<PeakHour[]> => {
    try {
      const response = await apiClient.get<PeakHour[]>(`${ANALYTICS_API}/peak-hours`);
      return response.data;
    } catch (error) {
      console.error('Error fetching peak hours:', error);
      throw error;
    }
  },

  /**
   * Get peak booking hours by date range
   */
  getPeakBookingHoursByDateRange: async (
    startDate: string,
    endDate: string
  ): Promise<PeakHour[]> => {
    try {
      const response = await apiClient.get<PeakHour[]>(`${ANALYTICS_API}/peak-hours/range`, {
        params: { startDate, endDate }
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching peak hours by date range:', error);
      throw error;
    }
  },

  /**
   * Get bookings by resource type
   */
  getBookingsByResourceType: async (): Promise<BookingsByType[]> => {
    try {
      const response = await apiClient.get<BookingsByType[]>(`${ANALYTICS_API}/by-type`);
      return response.data;
    } catch (error) {
      console.error('Error fetching bookings by type:', error);
      throw error;
    }
  },

  /**
   * Get resource utilization
   */
  getResourceUtilization: async (): Promise<TopResource[]> => {
    try {
      const response = await apiClient.get<TopResource[]>(`${ANALYTICS_API}/resource-utilization`);
      return response.data;
    } catch (error) {
      console.error('Error fetching resource utilization:', error);
      throw error;
    }
  }
};

File: backend/src/main/java/com/smartcampus/service/AnalyticsService.java
package com.smartcampus.service;

import com.smartcampus.dto.analytics.TopResourceDTO;
import com.smartcampus.dto.analytics.PeakHourDTO;
import com.smartcampus.dto.analytics.BookingsByTypeDTO;
import com.smartcampus.repository.BookingRepository;
import com.smartcampus.repository.UserRepository;
import com.smartcampus.repository.TicketRepository;
import com.smartcampus.repository.ResourceRepository;
import com.smartcampus.dto.AdminStatsDTO;
import com.smartcampus.model.Ticket;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

/**
 * Analytics Service
 * Provides aggregated analytics data for admin dashboard
 */
@Service
@Transactional(readOnly = true)
public class AnalyticsService {

    @Autowired
    private BookingRepository bookingRepository;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private TicketRepository ticketRepository;

    @Autowired
    private ResourceRepository resourceRepository;

    /**
     * Get summary stats for admin dashboard
     */
    public AdminStatsDTO getAdminStats() {
        return AdminStatsDTO.builder()
                .totalUsers(userRepository.count())
                .pendingBookings(bookingRepository.countByStatus("PENDING"))
                .openTickets(ticketRepository.countByStatus(Ticket.TicketStatus.OPEN))
                .activeResources(resourceRepository.count())
                .build();
    }

    /**
     * Get top N resources by booking count
     */
    public List<TopResourceDTO> getTopResources(int limit) {
        List<Object[]> results = bookingRepository.findTopResourcesByBookingCount();
        return convertToTopResourceDTO(results).stream()
                .limit(limit)
                .toList();
    }

    /**
     * Get top N resources by booking count within date range
     */
    public List<TopResourceDTO> getTopResourcesByDateRange(LocalDate startDate, LocalDate endDate, int limit) {
        List<Object[]> results = bookingRepository.findTopResourcesByBookingCountInRange(startDate, endDate);
        return convertToTopResourceDTO(results).stream()
                .limit(limit)
                .toList();
    }

    /**
     * Get peak booking hours (which hours are most booked)
     */
    public List<PeakHourDTO> getPeakBookingHours() {
        List<Object[]> results = bookingRepository.findPeakBookingHours();
        return convertToPeakHourDTO(results);
    }

    /**
     * Get peak booking hours within date range
     */
    public List<PeakHourDTO> getPeakBookingHoursByDateRange(LocalDate startDate, LocalDate endDate) {
        List<Object[]> results = bookingRepository.findPeakBookingHoursInRange(startDate, endDate);
        return convertToPeakHourDTO(results);
    }

    /**
     * Get bookings by resource type
     */
    public List<BookingsByTypeDTO> getBookingsByResourceType() {
        List<Object[]> results = bookingRepository.findBookingsByResourceType();
        List<BookingsByTypeDTO> dtos = new ArrayList<>();
        for (Object[] row : results) {
            BookingsByTypeDTO dto = new BookingsByTypeDTO();
            dto.setType((String) row[0]);
            dto.setBookingCount(((Number) row[1]).longValue());
            dtos.add(dto);
        }
        return dtos;
    }

    /**
     * Get resource utilization metrics
     */
    public List<TopResourceDTO> getResourceUtilization() {
        List<Object[]> results = bookingRepository.findResourceUtilization();
        List<TopResourceDTO> dtos = new ArrayList<>();
        for (Object[] row : results) {
            TopResourceDTO dto = new TopResourceDTO();
            dto.setResourceId(((Number) row[0]).longValue());
            dto.setResourceName((String) row[1]);
            dto.setBookingCount(((Number) row[3]).longValue());
            dtos.add(dto);
        }
        return dtos;
    }

    private List<TopResourceDTO> convertToTopResourceDTO(List<Object[]> results) {
        List<TopResourceDTO> dtos = new ArrayList<>();
        for (Object[] row : results) {
            TopResourceDTO dto = new TopResourceDTO();
            dto.setResourceId(((Number) row[0]).longValue());
            dto.setResourceName((String) row[1]);
            dto.setResourceType((String) row[2]);
            dto.setLocation((String) row[3]);
            dto.setBookingCount(((Number) row[4]).longValue());
            dtos.add(dto);
        }
        return dtos;
    }

    private List<PeakHourDTO> convertToPeakHourDTO(List<Object[]> results) {
        List<PeakHourDTO> dtos = new ArrayList<>();
        for (Object[] row : results) {
            PeakHourDTO dto = new PeakHourDTO();
            dto.setHour(((Number) row[0]).intValue());
            dto.setBookingCount(((Number) row[1]).longValue());
            dto.setTimeLabel(String.format("%02d:00 - %02d:00", dto.getHour(), dto.getHour() + 1));
            dtos.add(dto);
        }
        return dtos;
    }
}
